# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/malaikasaleem944/malaika_flyrank/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

For this lane, one row represents one content item for one client on one report date.

The main analysis window for this task is the mid-panel month of March 2026 (`month=2026-03`). The warehouse contains daily performance history from 2025-01-27 through 2026-06-30, but I will use March 2026 for feature development and verification rather than the final June 2026 sample.

The decision is made at the content-item level: which content should an editor prioritize for review or refresh?

In [78]:
import os

print("HF_TOKEN available:", os.environ.get("HF_TOKEN") is not None)

HF_TOKEN available: True


In [79]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "duckdb"
])

0

In [80]:
import duckdb
print("DuckDB version:", duckdb.__version__)

DuckDB version: 1.5.5


In [81]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "duckdb"
])

0

In [82]:
import duckdb
print("DuckDB version:", duckdb.__version__)

DuckDB version: 1.5.5


In [83]:
con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)

""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [84]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("HF_TOKEN available:", HF_TOKEN is not None)


HF_TOKEN available: True


In [85]:
import duckdb

print("DuckDB version:", duckdb.__version__)

DuckDB version: 1.5.5


In [86]:
con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [87]:
import sys
print(sys.executable)

c:\Users\Malaika Saleem\Documents\malaika_flyrank\.venv\Scripts\python.exe


In [88]:
import sys
import subprocess

cmd = [sys.executable, "-m", "pip", "install", "--upgrade", "pip"]
p = subprocess.Popen(cmd)
p.wait()

print("pip upgrade finished with code:", p.returncode)

pip upgrade finished with code: 0


In [89]:
import sys
import subprocess

cmd = [sys.executable, "-m", "pip", "install", "--upgrade", "duckdb"]
p = subprocess.Popen(cmd)
p.wait()

print("DuckDB installation finished with code:", p.returncode)

DuckDB installation finished with code: 0


In [90]:
import sys
import duckdb

print(sys.executable)
print("DuckDB:", duckdb.__version__)

c:\Users\Malaika Saleem\Documents\malaika_flyrank\.venv\Scripts\python.exe
DuckDB: 1.5.5


In [91]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("HF_TOKEN available:", HF_TOKEN is not None)

HF_TOKEN available: True


In [92]:
import duckdb

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [93]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("HF_TOKEN available:", HF_TOKEN is not None)
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)

HF_TOKEN available: True
Token length: 28


In [94]:
import duckdb

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

print("HF secret configured")

HF secret configured


In [95]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("HF_TOKEN available:", bool(HF_TOKEN))

HF_TOKEN available: True


In [96]:
import os
import duckdb

HF_TOKEN = os.environ.get("HF_TOKEN")

print("HF_TOKEN available:", bool(HF_TOKEN))
print("DuckDB version:", duckdb.__version__)

HF_TOKEN available: True
DuckDB version: 1.5.5


In [97]:
con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [98]:
import sys
print(sys.executable)

c:\Users\Malaika Saleem\Documents\malaika_flyrank\.venv\Scripts\python.exe


In [99]:
import sys
print(sys.executable)


c:\Users\Malaika Saleem\Documents\malaika_flyrank\.venv\Scripts\python.exe


In [100]:
import duckdb
import huggingface_hub
import os

print("DuckDB:", duckdb.__version__)
print("HF Hub:", huggingface_hub.__version__)
print("HF_TOKEN available:", bool(os.environ.get("HF_TOKEN")))

DuckDB: 1.5.5
HF Hub: 1.28.0
HF_TOKEN available: True


In [101]:
import os
import duckdb

HF_TOKEN = os.environ["HF_TOKEN"]

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [102]:
from pathlib import Path
import duckdb

march_file = Path.home() / ".cache" / "huggingface" / "hub" / "datasets--FlyRank--internship-warehouse" / "snapshots" / "50cbf7c3909d07be4d1b5906b4d09e882e5acbf2" / "fact_content_daily_performance" / "month=2026-03" / "data_0.parquet"

print(march_file)
print("Exists:", march_file.exists())

C:\Users\Malaika Saleem\.cache\huggingface\hub\datasets--FlyRank--internship-warehouse\snapshots\50cbf7c3909d07be4d1b5906b4d09e882e5acbf2\fact_content_daily_performance\month=2026-03\data_0.parquet
Exists: True


In [103]:
test = con.sql(f"""
SELECT *
FROM '{march_file}'
LIMIT 1
""").df()

display(test)

,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events,month
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,<NA>,20,0,67,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03


In [104]:
columns = con.sql(f"""
DESCRIBE SELECT *
FROM '{march_file}'
""").df()

columns[["column_name", "column_type"]]

,column_name,column_type
0,report_date,DATE
1,client_hash_id,VARCHAR
2,content_hash_id,VARCHAR
3,client_has_gsc,BOOLEAN
4,client_has_ga4,BOOLEAN
5,gsc_data_available,BOOLEAN
6,ga4_data_available,BOOLEAN
7,gsc_impressions,BIGINT
8,gsc_clicks,BIGINT
9,gsc_sum_position,BIGINT


In [105]:
grain_check = con.sql(f"""
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    COUNT(*) AS row_count
FROM '{march_file}'
GROUP BY
    report_date,
    client_hash_id,
    content_hash_id
HAVING COUNT(*) > 1
LIMIT 5
""").df()

grain_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,report_date,client_hash_id,content_hash_id,row_count


In [106]:
count_window = con.sql(f"""
SELECT
    COUNT(*) AS row_count,
    MIN(report_date) AS min_date,
    MAX(report_date) AS max_date
FROM '{march_file}'
""").df()

count_window

,row_count,min_date,max_date
0,9841378,2026-03-01,2026-03-31


In [107]:
count_window = con.sql(f"""
SELECT
    COUNT(*) AS row_count,
    MIN(report_date) AS min_date,
    MAX(report_date) AS max_date
FROM '{march_file}'
""").df()

count_window

,row_count,min_date,max_date
0,9841378,2026-03-01,2026-03-31


In [108]:
availability_check = con.sql(f"""
SELECT
    COUNT(*) AS rows_with_gsc_available
FROM '{march_file}'
WHERE gsc_data_available IS TRUE
""").df()

availability_check

,rows_with_gsc_available
0,3611061


## 2. Fields: feature / label / context / excluded

### Features

The final feature frame will contain at most five features. I will use only information that is available at the decision moment and is measured before the future outcome window.

Candidate historical performance features include:

- `gsc_impressions` — observed search visibility available for the decision-period observation.
- `gsc_clicks` — observed search clicks available before the outcome window.
- `gsc_avg_position` — observed search position available before the outcome window.
- `ga4_sessions` — observed analytics sessions available before the outcome window when GA4 data is available.
- One additional content-level feature from the appropriate warehouse dimension table, if it is available before the decision moment.

I will keep the final feature frame to exactly five features, as required by the assignment.

### Label / proxy

The target is a future content-performance outcome that indicates whether a content item requires attention.

The target must be calculated from information after the decision moment. It will not be used as a feature.
### Context

These fields identify or locate observations but are not model features:

- `client_hash_id` — identifies the client for grouping and splitting.
- `content_hash_id` — identifies the content item.
- `report_date` — identifies the observation date and defines the time window.

### Excluded

- `trend_direction` — excluded because it is an outcome-derived field and can leak the target.
- `trend_pct` — excluded because it is used to derive the decline outcome and therefore contains outcome information.
- Final-month (`2026-06`) data — excluded from development because it is the sealed final month.
- Any information measured after the decision moment — excluded to prevent future-information leakage.

## 3. Verify it with queries (grain, counts, missing values, windows)

I will verify the warehouse grain, the March 2026 row count and date span, and data availability using three queries. I will use the mid-panel March 2026 partition for development and keep the final June 2026 month sealed.

In [109]:
grain_check = con.sql(f"""
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    COUNT(*) AS row_count
FROM '{march_file}'
GROUP BY
    report_date,
    client_hash_id,
    content_hash_id
HAVING COUNT(*) > 1
LIMIT 5
""").df()

grain_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,report_date,client_hash_id,content_hash_id,row_count


**Grain result:** The query returned zero duplicate `(report_date, client_id, content_id)` combinations. This verifies that one row represents one content item for one client on one report date in the March 2026 partition.

In [110]:
count_window = con.sql(f"""
SELECT
    COUNT(*) AS row_count,
    MIN(report_date) AS min_date,
    MAX(report_date) AS max_date
FROM '{march_file}'
""").df()

count_window

,row_count,min_date,max_date
0,9841378,2026-03-01,2026-03-31


In [111]:
availability_check = con.sql(f"""
SELECT
    COUNT(*) AS rows_with_gsc_available
FROM '{march_file}'
WHERE gsc_data_available IS TRUE
""").df()

availability_check

,rows_with_gsc_available
0,3611061


In [112]:
grain_check = con.sql(f"""
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    COUNT(*) AS row_count
FROM '{march_file}'
GROUP BY
    report_date,
    client_hash_id,
    content_hash_id
HAVING COUNT(*) > 1
LIMIT 5
""").df()

grain_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,report_date,client_hash_id,content_hash_id,row_count


In [113]:
feature_df = con.sql(f"""
SELECT
    client_hash_id,
    content_hash_id,
    report_date,
    gsc_impressions,
    gsc_clicks,
    gsc_avg_position,
    ga4_sessions,
    scroll_events
FROM '{march_file}'
WHERE gsc_data_available IS TRUE
LIMIT 10000
""").df()

feature_df.head()

,client_hash_id,content_hash_id,report_date,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_sessions,scroll_events
0,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,2026-03-01,20,0,3.350000,<NA>,<NA>
1,client_73cda7b4e4f265ea,content_05597932fe4da067,2026-03-01,1,0,0.000000,<NA>,<NA>
2,client_73cda7b4e4f265ea,content_7a105f548d9c6916,2026-03-01,125,1,4.928000,<NA>,<NA>
3,client_73cda7b4e4f265ea,content_905aa32a0230694e,2026-03-01,7,0,4.000000,<NA>,<NA>
4,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,2026-03-01,11,0,2.272727,<NA>,<NA>


### Five features and availability

- `gsc_impressions` — available at the decision moment because it is observed search visibility from the current historical period.
- `gsc_clicks` — available at the decision moment because these are already observed search clicks.
- `gsc_avg_position` — available at the decision moment because it is measured search position from the historical period.
- `ga4_sessions` — available at the decision moment when `ga4_data_available IS TRUE`, because these sessions have already been observed.
- `scroll_events` — available at the decision moment because these engagement events have already been observed.

## 4. Data limits

This data cannot provide a perfectly balanced history for every client and content item because clients have different amounts of available search and analytics history.

Some rows have Google Analytics fields that are zero-filled when `ga4_data_available` is not true. Therefore, a zero value does not always mean that there was no engagement.

The fixed 90-day query window can also overlap with the performance snapshot window. Therefore, features and labels must be aligned carefully so that information from the future outcome window does not enter the features.

The final June 2026 month is treated as a sealed outcome/test period rather than being used to develop the label or features.

Therefore, the model can support content-prioritization decisions, but it cannot prove that a content change caused a future performance improvement.

In [114]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

leak_df = con.sql(f"""
SELECT
    client_hash_id,
    content_hash_id,

    -- Features: known before the future period
    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN gsc_impressions
            ELSE 0
        END
    ) AS gsc_impressions,

    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN gsc_clicks
            ELSE 0
        END
    ) AS gsc_clicks,

    AVG(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN gsc_avg_position
        END
    ) AS gsc_avg_position,

    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN ga4_sessions
            ELSE 0
        END
    ) AS ga4_sessions,

    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN scroll_events
            ELSE 0
        END
    ) AS scroll_events,

    -- Future outcome: second half of March
    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-16' AND DATE '2026-03-31'
            THEN gsc_clicks
            ELSE 0
        END
    ) AS future_clicks

FROM '{march_file}'
GROUP BY client_hash_id, content_hash_id
HAVING
    SUM(
        CASE
            WHEN report_date BETWEEN DATE '2026-03-01' AND DATE '2026-03-15'
            THEN gsc_impressions
            ELSE 0
        END
    ) > 0
""").df()

leak_df.head()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,client_hash_id,content_hash_id,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_sessions,scroll_events,future_clicks
0,client_62f4a7e64f5e0096,content_aa184c1b4ea518e5,133.0,0.0,10.836170,0.0,0.0,0.0
1,client_62f4a7e64f5e0096,content_7b1b1490a8e36a48,3.0,0.0,5.000000,0.0,0.0,0.0
2,client_62f4a7e64f5e0096,content_12d1c050115b68a7,2189.0,10.0,3.975472,0.0,0.0,4.0
3,client_62f4a7e64f5e0096,content_7addcc068ecd8213,445.0,0.0,5.438703,0.0,0.0,0.0
4,client_62f4a7e64f5e0096,content_62a37f1ef4dcf407,1052.0,1.0,3.932680,0.0,0.0,0.0


### Deliberate leakage experiment

I deliberately added `leaked_label`, which directly contains the target label. The resulting score increased toward perfect performance because the model was given the answer it was supposed to predict.

I then removed `leaked_label` and evaluated the model using only the five decision-time features. The lower honest score is the number that should be retained.

This demonstrates that a very high score can be caused by target leakage rather than genuine predictive ability.

In [115]:
import sklearn
print("scikit-learn:", sklearn.__version__)

scikit-learn: 1.9.0


In [116]:
print(leak_df.columns.tolist())

['client_hash_id', 'content_hash_id', 'gsc_impressions', 'gsc_clicks', 'gsc_avg_position', 'ga4_sessions', 'scroll_events', 'future_clicks']


In [117]:
median_future_clicks = leak_df["future_clicks"].median()

leak_df["label"] = (
    leak_df["future_clicks"] < median_future_clicks
).astype(int)

print("Label created successfully")
print(leak_df[["future_clicks", "label"]].head())

Label created successfully
   future_clicks  label
0            0.0      0
1            0.0      0
2            4.0      0
3            0.0      0
4            0.0      0


In [118]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Deliberately leak the target into the features
leak_df["leaked_label"] = leak_df["label"]

X = leak_df[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_sessions",
        "scroll_events",
        "leaked_label"
    ]
]

y = leak_df["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

model = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

model.fit(X_train, y_train)

pred = model.predict(X_test)

leaked_score = accuracy_score(y_test, pred)

print("Accuracy with deliberate leakage:", leaked_score)

Accuracy with deliberate leakage: 1.0


In [119]:
# Remove the deliberately leaked column
honest_df = leak_df.drop(columns=["leaked_label"])

# Use only the five legitimate features
X_honest = honest_df[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_sessions",
        "scroll_events"
    ]
]

y_honest = honest_df["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X_honest,
    y_honest,
    test_size=0.2,
    random_state=42,
    stratify=y_honest
)

honest_model = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

honest_model.fit(X_train, y_train)

honest_pred = honest_model.predict(X_test)

honest_score = accuracy_score(
    y_test,
    honest_pred
)

print("Honest accuracy after removing leakage:", honest_score)

Honest accuracy after removing leakage: 1.0


In [120]:
# Create a balanced-enough binary outcome based on future clicks
leak_df["label"] = (
    leak_df["future_clicks"] > 0
).astype(int)

print("New label distribution:")
print(leak_df["label"].value_counts())

print("\nNew label proportions:")
print(leak_df["label"].value_counts(normalize=True))

New label distribution:
label
0    101118
1     50863
Name: count, dtype: int64

New label proportions:
label
0    0.665333
1    0.334667
Name: proportion, dtype: float64


In [121]:
print("Median future clicks:", median_future_clicks)
print("Total rows:", len(leak_df))

Median future clicks: 0.0
Total rows: 151981


In [122]:
# Deliberately add the target as a feature
leak_df["leaked_label"] = leak_df["label"]

X = leak_df[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_sessions",
        "scroll_events",
        "leaked_label"
    ]
]

y = leak_df["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

leak_model = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

leak_model.fit(X_train, y_train)

leak_pred = leak_model.predict(X_test)

leaked_score = accuracy_score(y_test, leak_pred)

print("Accuracy with deliberate leakage:", leaked_score)

Accuracy with deliberate leakage: 1.0


In [123]:
# Remove the leaked target
honest_df = leak_df.drop(columns=["leaked_label"])

X_honest = honest_df[
    [
        "gsc_impressions",
        "gsc_clicks",
        "gsc_avg_position",
        "ga4_sessions",
        "scroll_events"
    ]
]

y_honest = honest_df["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X_honest,
    y_honest,
    test_size=0.2,
    random_state=42,
    stratify=y_honest
)

honest_model = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

honest_model.fit(X_train, y_train)

honest_pred = honest_model.predict(X_test)

honest_score = accuracy_score(y_test, honest_pred)

print("Honest accuracy after removing leakage:", honest_score)

Honest accuracy after removing leakage: 0.8405434746850018


### Leakage lesson

The deliberate leakage experiment produced an accuracy of 1.00 because `leaked_label` directly contained the target label.

After removing `leaked_label`, the accuracy decreased to 0.841 (84.07%). This is the honest result using only the five decision-time features.

The difference shows why label-derived information must never be included as a model feature. A near-perfect score can be caused by leakage rather than genuine predictive performance.

In [124]:
print("Final features:")
print(list(X_honest.columns))

Final features:
['gsc_impressions', 'gsc_clicks', 'gsc_avg_position', 'ga4_sessions', 'scroll_events']


## Self-check

- [ ] Every section above is filled with both the markdown explanation and the code that supports it.
- [ ] The notebook runs from top to bottom with no errors after Runtime → Run all.
- [ ] No client names, private URLs, access tokens, or private queries are included.
- [ ] My claims use careful language such as observed, measured, directional, or decision-support.
- [ ] The deliberate leakage experiment is shown, and the leaked feature is removed from the final feature frame.
- [ ] The notebook is committed under `work/notebooks/` and pushed to my repository.